File: ai_diagnostic/batch.py
import csv
import json
from collections import Counter
from typing import Optional

from .evidence import Evidence, EVIDENCE_PARAMS, evidence_from_dict
from .bayesian import BayesianDiagnostic, DiagnosisResult
from .likelihoods import FAILURE_CLASSES, FAILURE_CLASS_LABELS


class BatchProcessor:
    def __init__(self, engine: Optional[BayesianDiagnostic] = None):
        self.engine = engine or BayesianDiagnostic()

    def process(self, evidence_list: list[Evidence], priors: dict = None) -> list[DiagnosisResult]:
        return [
            self.engine.diagnose(evidence, prior_overrides=priors)
            for evidence in evidence_list
        ]

    def process_from_dicts(self, dicts: list[dict]) -> list[DiagnosisResult]:
        return self.process([evidence_from_dict(d) for d in dicts])

    def summary(self, results: list[DiagnosisResult]) -> dict:
        if not results:
            return {
                "total": 0,
                "by_type": {},
                "average_confidence": 0.0,
                "flagged_for_investigation": 0,
                "max_confidence": 0.0,
                "min_confidence": 0.0,
            }

        type_counts = Counter(r.diagnosis for r in results)
        confidences = [r.confidence for r in results]
        flagged = sum(1 for r in results if r.needs_investigation)

        return {
            "total": len(results),
            "by_type": {
                FAILURE_CLASS_LABELS.get(k, k): v
                for k, v in type_counts.most_common()
            },
            "average_confidence": round(sum(confidences) / len(confidences), 4),
            "max_confidence": round(max(confidences), 4),
            "min_confidence": round(min(confidences), 4),
            "flagged_for_investigation": flagged,
        }

    def compare(self, results: list[DiagnosisResult]) -> str:
        if not results:
            return ""

        headers = ["Case", "Diagnosis", "Confidence", "Investigate?"]
        for cause in FAILURE_CLASSES:
            label = FAILURE_CLASS_LABELS.get(cause, cause)
            headers.append(label)

        rows = []
        for i, r in enumerate(results, 1):
            row = [
                str(i),
                FAILURE_CLASS_LABELS.get(r.diagnosis, r.diagnosis),
                f"{r.confidence:.2%}",
                "Yes" if r.needs_investigation else "No",
            ]
            for cause in FAILURE_CLASSES:
                row.append(f"{r.posteriors.get(cause, 0.0):.4f}")
            rows.append(row)

        col_widths = [max(len(h), max((len(row[j]) for row in rows), default=0))
                      for j, h in enumerate(headers)]

        def fmt_row(vals):
            return "  ".join(v.ljust(w) for v, w in zip(vals, col_widths))

        lines = [
            fmt_row(headers),
            "  ".join("-" * w for w in col_widths),
        ]
        for row in rows:
            lines.append(fmt_row(row))
        return "\n".join(lines)

    def export_batch(self, results: list[DiagnosisResult], path: str, format: str = "csv"):
        if format == "json":
            data = []
            for i, r in enumerate(results, 1):
                entry = {
                    "case": i,
                    "diagnosis": r.diagnosis,
                    "diagnosis_label": FAILURE_CLASS_LABELS.get(r.diagnosis, r.diagnosis),
                    "confidence": round(r.confidence, 4),
                    "needs_investigation": r.needs_investigation,
                    "posteriors": {k: round(v, 4) for k, v in r.posteriors.items()},
                }
                data.append(entry)
            with open(path, "w") as f:
                json.dump(data, f, indent=2)
        else:
            with open(path, "w", newline="") as f:
                writer = csv.writer(f)
                header = ["case", "diagnosis", "diagnosis_label", "confidence", "needs_investigation"]
                header += [FAILURE_CLASS_LABELS.get(c, c) for c in FAILURE_CLASSES]
                writer.writerow(header)
                for i, r in enumerate(results, 1):
                    row = [
                        i,
                        r.diagnosis,
                        FAILURE_CLASS_LABELS.get(r.diagnosis, r.diagnosis),
                        f"{r.confidence:.4f}",
                        r.needs_investigation,
                    ]
                    row += [f"{r.posteriors.get(c, 0.0):.4f}" for c in FAILURE_CLASSES]
                    writer.writerow(row)

File: ai_diagnostic_rs/src/batch.rs
//! Batch processing, summaries, side-by-side comparison tables, exports.

use crate::bayesian::{BayesianDiagnostic, DiagnosisResult};
use crate::evidence::{Evidence, evidence_from_map};
use crate::likelihoods::{failure_class_label, FAILURE_CLASSES};
use serde_json::{json, Map, Value};
use std::collections::HashMap;

pub struct BatchProcessor {
    pub engine: BayesianDiagnostic,
}

impl Default for BatchProcessor {
    fn default() -> Self {
        Self::new()
    }
}

impl BatchProcessor {
    pub fn new() -> Self {
        BatchProcessor {
            engine: BayesianDiagnostic::new(),
        }
    }

    pub fn with_engine(engine: BayesianDiagnostic) -> Self {
        BatchProcessor { engine }
    }

    pub fn process(
        &self,
        evidence_list: &[Evidence],
        priors: Option<&HashMap<String, f64>>,
    ) -> Result<Vec<DiagnosisResult>, String> {
        evidence_list
            .iter()
            .map(|e| self.engine.diagnose(e, priors, 0.6))
            .collect()
    }

    pub fn process_from_dicts(
        &self,
        dicts: &[Map<String, Value>],
    ) -> Result<Vec<DiagnosisResult>, String> {
        let mut all = Vec::with_capacity(dicts.len());
        for d in dicts {
            let ev = evidence_from_map(d)?;
            all.push(self.engine.diagnose(&ev, None, 0.6)?);
        }
        Ok(all)
    }

    pub fn summary(&self, results: &[DiagnosisResult]) -> Value {
        if results.is_empty() {
            return json!({
                "total": 0,
                "by_type": {},
                "average_confidence": 0.0,
                "max_confidence": 0.0,
                "min_confidence": 0.0,
                "flagged_for_investigation": 0,
            });
        }

        let mut type_counts: Vec<(String, usize)> = Vec::new();
        for r in results {
            match type_counts.iter_mut().find(|(c, _)| c == &r.diagnosis) {
                Some((_, n)) => *n += 1,
                None => type_counts.push((r.diagnosis.clone(), 1)),
            }
        }
        type_counts.sort_by_key(|(_, n)| std::cmp::Reverse(*n));

        let confidences: Vec<f64> = results.iter().map(|r| r.confidence).collect();
        let avg = confidences.iter().sum::<f64>() / confidences.len() as f64;
        let round4 = |x: f64| (x * 10000.0).round() / 10000.0;

        json!({
            "total": results.len(),
            "by_type": type_counts
                .into_iter()
                .map(|(c, n)| (failure_class_label(&c).to_string(), json!(n)))
                .collect::<Map<String, Value>>(),
            "average_confidence": round4(avg),
            "max_confidence": round4(confidences.iter().cloned().fold(f64::MIN, f64::max)),
            "min_confidence": round4(confidences.iter().cloned().fold(f64::MAX, f64::min)),
            "flagged_for_investigation": results.iter().filter(|r| r.needs_investigation).count(),
        })
    }

    /// Side-by-side comparison table. `labels` optionally names each case.
    pub fn compare(
        &self,
        results: &[DiagnosisResult],
        labels: Option<&[String]>,
    ) -> String {
        if results.is_empty() {
            return String::new();
        }

        let mut headers: Vec<String> = vec![
            "Case".to_string(),
            "Diagnosis".to_string(),
            "Confidence".to_string(),
            "Investigate?".to_string(),
        ];
        for cause in FAILURE_CLASSES {
            headers.push(failure_class_label(cause).to_string());
        }

        let mut rows: Vec<Vec<String>> = Vec::new();
        for (i, r) in results.iter().enumerate() {
            let label = labels
                .and_then(|ls| ls.get(i))
                .cloned()
                .unwrap_or_else(|| i.to_string());
            let mut row = vec![
                label,
                failure_class_label(&r.diagnosis).to_string(),
                format!("{:.2}%", r.confidence * 100.0),
                if r.needs_investigation { "Yes" } else { "No" }.to_string(),
            ];
            for cause in FAILURE_CLASSES {
                row.push(format!("{:.4}", r.posterior(cause)));
            }
            rows.push(row);
        }

        let col_widths: Vec<usize> = headers
            .iter()
            .enumerate()
            .map(|(j, h)| {
                rows.iter()
                    .map(|row| row[j].chars().count())
                    .chain(std::iter::once(h.chars().count()))
                    .max()
                    .unwrap_or(0)
            })
            .collect();

        let fmt_row = |vals: &[String]| -> String {
            vals.iter()
                .zip(col_widths.iter())
                .map(|(v, w)| {
                    let pad = w.saturating_sub(v.chars().count());
                    format!("{v}{}", " ".repeat(pad))
                })
                .collect::<Vec<_>>()
                .join("  ")
        };

        let mut lines = vec![fmt_row(&headers)];
        lines.push(col_widths
            .iter()
            .map(|w| "-".repeat(*w))
            .collect::<Vec<_>>()
            .join("  "));
        for row in &rows {
            lines.push(fmt_row(row));
        }
        lines.join("\n")
    }

    pub fn export_batch(
        &self,
        results: &[DiagnosisResult],
        path: &std::path::Path,
        format: &str,
    ) -> Result<(), String> {
        match format {
            "json" => {
                let data: Vec<Value> = results
                    .iter()
                    .enumerate()
                    .map(|(i, r)| {
                        let round4 =
                            |x: f64| (x * 10000.0).round() / 10000.0;
                        json!({
                            "case": i + 1,
                            "diagnosis": r.diagnosis,
                            "diagnosis_label": failure_class_label(&r.diagnosis),
                            "confidence": round4(r.confidence),
                            "needs_investigation": r.needs_investigation,
                            "posteriors": r
                                .posteriors
                                .iter()
                                .map(|(k, v)| (k.clone(), json!(round4(*v))))
                                .collect::<Map<String, Value>>(),
                        })
                    })
                    .collect();
                serde_json::to_writer_pretty(std::fs::File::create(path).map_err(|e| e.to_string())?, &data)
                    .map_err(|e| e.to_string())
            }
            _ => {
                let mut w = csv::Writer::create(path).map_err(|e| e.to_string())?;
                let mut header: Vec<String> = vec![
                    "case".into(),
                    "diagnosis".into(),
                    "diagnosis_label".into(),
                    "confidence".into(),
                    "needs_investigation".into(),
                ];
                header.extend(FAILURE_CLASSES.map(|c| failure_class_label(c).to_string()));
                w.row(&header);
                for (i, r) in results.iter().enumerate() {
                    let mut row = vec![
                        (i + 1).to_string(),
                        r.diagnosis.clone(),
                        failure_class_label(&r.diagnosis).to_string(),
                        format!("{:.4}", r.confidence),
                        r.needs_investigation.to_string(),
                    ];
                    row.extend(FAILURE_CLASSES.map(|c| format!("{:.4}", r.posterior(c))));
                    w.row(&row);
                }
                w.flush().map_err(|e| e.to_string())
            }
        }
    }
}

mod csv {
    use std::fs::File;
    use std::io::{BufWriter, Write};
    use std::path::Path;

    pub struct Writer {
        inner: BufWriter<File>,
    }

    impl Writer {
        pub fn create(path: &Path) -> std::io::Result<Self> {
            Ok(Writer {
                inner: BufWriter::new(File::create(path)?),
            })
        }

        pub fn row(&mut self, fields: &[String]) {
            let line = fields
                .iter()
                .map(|f| {
                    if f.contains(',') || f.contains('"') || f.contains('\n') {
                        format!("\"{}\"", f.replace('"', "\"\""))
                    } else {
                        f.clone()
                    }
                })
                .collect::<Vec<_>>()
                .join(",");
            let _ = writeln!(self.inner, "{line}");
        }

        pub fn flush(mut self) -> std::io::Result<()> {
            self.inner.flush()
        }
    }
}
